### PR TITLE
Rich text: remove preserveWhiteSpace serialisation differences

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

This PR adjusts the `preserveWhiteSpace` prop from rich text, which was being used for the Preformatted, Code, and Verse blocks. We don't have to continuously collapse whitespace inside rich text on value changes, and we don't need it for serialisation. Additionally, it's no longer needed for splitting/merging all instances serialise the same way.

## Why?

1. This regex to remove whitespace is being continuously run on typing, so this impacts performance positively.
2. This setting (along with `multiline`, which has recently been removed) is a pain to work with. Currently, to parse and serialised rich text content (with `create` and `toHTMLString`), you need to know where this HTML came from (an instance that preserves white space or not). This required us to add a special setting to the block attribute definition, so we can properly handle merging and splitting. Now all rich text values serialise the same way, with line breaks always serialising to the HTML BR tag.

## How?

As said before, we can rely on the initial HTML parse and the paste handler to remove superfluous white space. The logic in the paste handler was introduces in #17470, much after we added this `preserveWhiteSpace` prop to rich text.

Additionally, for line break serialisation, it should be safe to use proper line break elements for _rich_ text instances. It's worth noting that Preformatted, Code, and Verse use _rich_ text components, allowing HTML formatting like bold, links, etc., so line break HTML elements are not odd. If we don't want any HTML elements, we should use the _plain_ text component instead, which outputs line breaks as simple text line breaks.

## Testing Instructions

We have lots of e2e test coverage, so we should be good. Mainly check the Preformatted, Code, and Verse blocks.

### CODE REVIEW

File: packages/block-editor/src/store/actions.js [hunk 1]
```diff
@@ -759,43 +758,23 @@ export const __unstableDeleteSelection =
...
-			blockBType.attributes[ selectionB.attributeKey ];
-
-		let valueA = create( {
-			html: htmlA,
-			...mapRichTextSettings( attributeDefinitionA ),
-		} );
-		let valueB = create( {
-			html: htmlB,
-			...mapRichTextSettings( attributeDefinitionB ),
-		} );
```
Comment (ellatrix) [inline packages/block-editor/src/store/actions.js:782]: Like I said, a pain to work with.

File: packages/rich-text/src/create.js [hunk 2]
```diff
@@ -268,16 +264,6 @@ function filterRange( node, range, filter ) {
...
 }
 
-/**
- * Collapse any whitespace used for HTML formatting to one space character,
- * because it will also be displayed as such by the browser.
- *
- * @param {string} string
- */
-function collapseWhiteSpace( string ) {
-	return string.replace( /[\n\r\t]+/g, ' ' );
```
Comment (ellatrix) [inline packages/rich-text/src/create.js:278]: This is sort of replaced by logic in the paste handler

https://github.com/WordPress/gutenberg/blob/acd54770808254a650e296f88d25a7e0da5e8596/packages/blocks/src/api/raw-handling/html-formatting-remover.js#L50

**Review by dmsnell — APPROVED** [review]
I've been testing this with the [Playground](https://playground.wordpress.net/#%7B%22landingPage%22:%22/wp-admin/%22,%22steps%22:%5B%7B%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22%7D,%7B%22step%22:%22mkdir%22,%22path%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22writeFile%22,%22path%22:%22/wordpress/pr/pr.zip%22,%22data%22:%7B%22resource%22:%22url%22,%22url%22:%22/plugin-proxy.php?org=WordPress&repo=gutenberg&workflow=Build%2520Gutenberg%2520Plugin%2520Zip&artifact=gutenberg-plugin&pr=55999%22,%22caption%22:%22Downloading%20Gutenberg%20PR%2055999%22%7D,%22progress%22:%7B%22weight%22:2,%22caption%22:%22Applying%20Gutenberg%20PR%2055999%22%7D%7D,%7B%22step%22:%22unzip%22,%22zipPath%22:%22/wordpress/pr/pr.zip%22,%22extractToPath%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22installPlugin%22,%22pluginZipFile%22:%7B%22resource%22:%22vfs%22,%22path%22:%22/wordpress/pr/gutenberg.zip%22%7D%7D%5D%7D) and can't find a way to break it.

I explored variations of whitepspace and conversion to and from a paragraph, code block, HTML block.

The only thing that stood out to me was that my spaces were preserved inside the editor for the paragraph but obviously didn't remain in the rendered post. I would have expected successive spaces in the editor (for example, typing `this      alignment`) would produce HTML that preserves those spaces, such as with a non-breaking space. Since this happens in `trunk` it's not a blocker or even a comment about this change specifically.

While reviewing I did wonder what our reason has been for removing whitespace at all. If we don't preserve the _intent_ or entered text from the editor to the rendered page, why modify it at all? Exploring the DOM inside the browser revealed that the browser doesn't modify `textContent` to remove extra whitespace, and I couldn't find any way to make it do it, even with calls to `.normalize()` and `.cloneNode()`. In other words, if browsers handle the whitespace and the editor has is own rules, what are we trying to accomplish by messing with whitespace?

**Comment by ellatrix** [comment]
That's because we have white space: pre-wrap turned on for all rich text instances. I'll try to remove it in [587dede](https://github.com/WordPress/gutenberg/pull/55999/commits/587dede9eead4f2b639a6a23b2a52f9697f65819).

**Comment by westonruter** [comment]
FYI: This change apparently broke my Syntax-highlighting Code Block plugin, since the `<pre>` contents no longer contain newlines but rather `<br>` tags: https://github.com/westonruter/syntax-highlighting-code-block/issues/790

I have a [workaround](https://github.com/westonruter/syntax-highlighting-code-block/pull/791) to replace all `<br>` with `\n` before doing the server-side highlighting.

**Comment by ellatrix** [comment]
@westonruter Worth nothing that the RichText and the code block can contain any HTML formatting. If you create a custom block and don't want formatting, you should ideally use the PlainText component (v2).

**Comment by westonruter** [comment]
@ellatrix Thanks. Good point. However, I'm hoping to retain the ability to do HTML formatting _and_ have syntax highlighting (https://github.com/westonruter/syntax-highlighting-code-block/issues/205), likely facilitated with the HTML Tag Processor. If I fail at that, I'll go with what you suggest.

**Comment by dmsnell** [comment]
> This change apparently broke my Syntax-highlighting Code Block plugin, since the `<pre>` contents no longer contain newlines but rather `<br>` tags

@westonruter are you looking to render the contents of the PRE element verbatim as code? if so, I'm not sure that's a reliable method anyway. the PRE element doesn't add any semantic to its contents other than that they should be monospace-rendered.

have you looked at storing the contents in a disabled TEXTAREA? or have you tried first decoding the contents on the server? if we interpret the PRE contents appropriately I wonder if the problem would persist? that's now doable with the HTML API in 6.5. does the syntax highlighter have the same struggles with encoded characters?

**Pull request merged**